### PR TITLE
doc/user: document refactor to controller-runtime

### PR DESCRIPTION
**Pull request by estroz**
The refactor from SDK-specific k8s API calls to `kubernetes-sigs/controller-runtime` was
a large undertaking that changed project structure and how generated projects interact
with a k8s cluster. This document attempts to explain these changes to both new users
and developers of existing operators generated with SDK `<= v0.7.0`.

### CODE REVIEW

**Comment by hasbro17** [comment]
@estroz I was thinking of this document less as a proposal and more of a user facing document demonstrating the controller-runtime's client API and example usage.

So we can focus more on just the client in this doc. And less on the manager.
We can move this from `doc/design/...` to `doc/user/client.md` or something similar.

And restructure the sections `Background` and `Proposal` into an `Overview` section that goes over how the client is made available to be used in the reconciler in a default controller. And how that client is backed by the cache for reads.
Since we're not instructing them to create a new manager we probably don't need the section on how to create a new manager.

We'll probably have a separate doc for the manager later.

File: doc/user/client.md [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	
+	...
+)
+
+func main() {
+	...
+
+	cfg, _ := config.GetConfig()
+	mgr, _ := manager.New(cfg, manager.Options{Namespace: namespace})
```
Comment (hasbro17) [inline doc/user/client.md]: You shouldn't need to ever change the manager's client.

This section should just demonstrate how to create a new client object if you're not using the Reconciler's client for some reason. (Say in a utility function somewhere).
```Go
import (
    "sigs.k8s.io/controller-runtime/pkg/client"
)

cfg, _ := config.GetConfig()
client, _ := client.New(cfg, client.Options{})
```

File: doc/user/client.md [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+
+controller-runtime defines several interfaces used for cluster interaction:
+- `client.Client`: implementers perform CRUD operations on a k8s cluster.
+- `manager.Manager`: manages shared dependencies, such as Caches and Clients.
+- `reconcile.Reconciler`: compares provided state with actual cluster state and updates the cluster on finding state differences using a Client.
+
+The SDK relies on a `manager.Manager` to create a split `client.Client` interface that performs Create, Update, Delete, Get, and List operations within a `reconcile.Reconciler`'s Reconcile function.
```
Comment (hasbro17) [inline doc/user/client.md]: This line is specific to the default Reconciler client populated by the manger. So you can move it down and consolidate it into the `Default Client` section.

File: doc/user/client.md [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+
+## Client Usage
+
+### Default Client
+
+The SDK will generate code creating a Manager. Managers hold a Cache and a Client that are used in CRUD operations and interface with the API server. The particular Client the SDK uses by default is referred to as a [split client][doc-split-client], created when instantiating a Manager. A split client reads (Get and List) from the Cache and writes (Create, Update, Delete) to the API server. Reading from the Cache significantly reduces request load on the API server; as long as the Cache is updated by the API server, read operations are eventually consistent.
```
Comment (hasbro17) [inline doc/user/client.md]: `The particular Client the SDK uses by default is referred to as a [split client][doc-split-client], created when instantiating a Manager.`

The SDK isn't really providing or using the client. So we can rephrase this to something like:

`By default a Controller's Reconciler will be populated with the Manager's client which is a split-client[doc-split-client].`
```GO
func newReconciler(mgr manager.Manager) reconcile.Reconciler {
	return &ReconcileMemcached{client: mgr.GetClient(), scheme: mgr.GetScheme()}
}

type ReconcileMemcached struct {
	client client.Client
	scheme *runtime.Scheme
}
```

File: doc/user/client.md (hunk 1 above)
Comment (estroz) [reply doc/user/client.md]: So instead of calling
```Go
func (r *ReconcileKind) Reconcile(...) {
    r.client.Get(...)
}
```

You would create a new client and use it instead:
```Go
func (r *ReconcileKind) Reconcile(...) {
    cfg, _ := config.GetConfig()
    client, _ := client.New(cfg, client.Options{})
    client.Get(...)
}
```

Comment (hasbro17) [reply doc/user/client.md]: Kind of. Although you shouldn't really be creating a new client in the reconcile loop.
That should be done in the constructor for your reconciler:
```Go
func newReconciler(mgr manager.Manager) reconcile.Reconciler {
        cfg, _ := config.GetConfig()
        client, _ := client.New(cfg, client.Options{Scheme: mgr.GetScheme()})
	return &ReconcileMemcached{client: client, scheme: mgr.GetScheme()}
}

type ReconcileMemcached struct {
	client client.Client
	scheme *runtime.Scheme
}
```
But even creating a new client in your reconciler isn't advised unless you know what you're doing. 
You'd need to set the scheme correctly as well so that client recognizes your custom types.

That's why I didn't specify where to create a new client so users don't get the idea to create a new client in their reconciler.
This is more for if you're writing a test and you need to instantiate a client for that. Or in some utility package.

Comment (estroz) [reply doc/user/client.md]: I'll remove the specifics.

File: doc/user/client.md [hunk 4]
```diff
@@ -0,0 +1,252 @@
+# Operator-SDK: Controller Runtime Client API
+
+## Overview
+
+The [kubernetes-sigs][org-kubernetes-sigs] organization builds tools to call the [Kubernetes][site-kubernetes] API in a clean, abstracted manner, much like the Operator SDK. The [controller-runtime][repo-controller-runtime] project is meant to help users build [controllers][site-kubernetes-controllers] easily by generating a controller project that interacts with a k8s cluster via CRUD operations. User-defined controller can perform specific tasks in a cluster; [operators][site-operators] can use multiple controllers to perform a variety of tasks. As operators use at least one controller, the SDK can rely on controller-runtime's k8s API code rather than develop a parallel set of API calls to execute the same cluster operations, namely CRUD operations.
```
Comment (hasbro17) [inline doc/user/client.md]: How about making this section of the overview a bit more concise and specific to the controller-runtime's Client. 

Right now this is giving a broad overview of several things plus it's going more into the motivation of why the SDK uses the controller-runtime. That's explanation is more suited for a proposal.

How about something like:
```
The controller-runtime library provides various abstractions to build controllers to watch and reconcile resources.
One of these abstractions is a generic client interface that let's a controller perform CRUD operations for different resource types.
```

File: doc/user/client.md [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+	client client.Client
+	scheme *runtime.Scheme
+}
+```
+
+A split client reads (Get and List) from the Cache and writes (Create, Update, Delete) to the API server. Reading from the Cache significantly reduces request load on the API server; as long as the Cache is updated by the API server, read operations are eventually consistent. 
+
+### Non-default Client
+
+An operator developer may wish to create their own Client with different read/write optimzations, for example. controller-runtime provides a [constructor][doc-client-constr] for Clients:
```
Comment (hasbro17) [inline doc/user/client.md]: We can be more specific here:
`with different read/write optimzations`==>`that serves read requests(Get List) from the apiserver instead of the cache`

File: doc/user/client.md [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+
+```Go
+// New returns a new Client using the provided config and Options.
+func New(config *rest.Config, options client.Options) (client.Client, error)
```
Comment (hasbro17) [inline doc/user/client.md]: Illustrate the options here:
https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/client/client.go#L34-L41

And add a line mentioning that the default scheme will have the core k8s resource types registered.
https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/client/client.go#L51
But for the client to recognize custom types users must provide a scheme that has their custom types registered.

File: doc/user/client.md [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+```
+
+Creating a new Client is not usually necessary nor advised, as the default Client is sufficient for most use cases. One reason to create a custom Client is for writing tests, a situation that might require a mock Client. controller-runtime provides a [fake Client][doc-client-fake] package for this purpose.
```
Comment (shawn-hurley) [inline doc/user/client.md]: > controller-runtime provides a [fake Client][doc-client-fake] package for this purpose.

Might be nice to also have an example of this here.

File: doc/user/client.md [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+[org-kubernetes-sigs]:https://github.com/kubernetes-sigs
+[site-kubernetes]:https://kubernetes.io/
+[site-kubernetes-controllers]:https://kubernetes.io/docs/concepts/workloads/controllers/
+[site-operators]:https://coreos.com/blog/introducing-operator-framework
+[doc-split-client]:https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/client/split.go#L26-L28
+[doc-rest-config]:https://godoc.org/k8s.io/client-go/rest#Config
+[doc-client-constr]:https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/client/client.go#L44
+[doc-client-fake]:https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/client/fake/client.go
+[doc-client-options]:https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/client/client.go#L35
+[doc-reconcile-reconciler]:https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/reconcile/reconcile.go#L35
```
Comment (shawn-hurley) [inline doc/user/client.md]: For some of the links, I think it might be better to point to the [kubebuilder book](https://book.kubebuilder.io/) or the [godoc](https://godoc.org/sigs.k8s.io/controller-runtime/pkg). These might be a better place to point users to get info than the code IMO

File: doc/user/client.md [hunk 9]
```diff
@@ -0,0 +1,252 @@
...
+func New(config *rest.Config, options client.Options) (client.Client, error)
+```
+
+Creating a new Client is not usually necessary nor advised, as the default Client is sufficient for most use cases. One reason to create a custom Client is for writing tests, a situation that might require a mock Client. controller-runtime provides a [fake Client][doc-client-fake] package for this purpose.
+
+### Reconcile and the Client API
+
+A Reconciler implements the [`reconcile.Reconciler`][doc-reconcile-reconciler] interface, which exposes the Reconcile method. Reconcilers are added to a corresponding Controller for a Kind; Reconcile is called in response to cluster or external Events, with a `reconcile.Request` object argument, to read and write cluster state by the Controller, and returns a `reconcile.Result`. SDK Reconcilers have access to a Client in order to make k8s API calls.
+
+**Note**: For those familiar with the SDK's old project structure, Reconcile replaces [Handle][doc-osdk-handle].
```
Comment (shawn-hurley) [inline doc/user/client.md]: I think we might want to be a little more pedantic here. The reconciler will replace the handle function of a single object/struct/gvk. The old handler you could have `v1alpha1.MyResource` and `v1alpha1.MySecondResource` handled by the same `handle` function.

File: doc/user/client.md (hunk 8 above)
Comment (hasbro17) [reply doc/user/client.md]: The runtime godoc might be better for each pkg. That's more readable.

File: doc/user/client.md (hunk 9 above)
Comment (hasbro17) [reply doc/user/client.md]: @shawn-hurley We're going to have a separate doc highlighting the major differences between the old SDK APIs and project layout vs the controller-runtime pkgs. And how sdk.Handle() maps to reconciler.Reconcile() etc.
That's another doc that I'm working on along with the migration doc.
We can point to that doc from here.

Comment (shawn-hurley) [reply doc/user/client.md]: I agree a more detailed doc is important, I guess I was trying to say that it might be better to be really explicit here. Assuming this will be merged before the migration doc, I don't want to give the wrong impression even if we clarify later. Thoughts?

Comment (hasbro17) [reply doc/user/client.md]: Yeah we can be a little more explicit here and mention that Reconcile handles the reconciliation requests for a single resource type unlike the old sdk handle function which could receive events for multiple resource types.

But probably not delve more into the specific interface and why that's so(since that gets into how you watch multiple resource in the controller). Those details can be deferred to the API differences doc.

Comment (estroz) [reply doc/user/client.md]: @hasbro17 @shawn-hurley does this seem reasonable now?

File: doc/user/client.md (hunk 7 above)
Comment (estroz) [reply doc/user/client.md]: We're removing the `fake` package reference and will address unit testing in a separate doc.

File: doc/user/client.md [hunk 10]
```diff
@@ -0,0 +1,411 @@
...
+
+	...
+
+	ctx := context.TODO()
+	if pod.Status.Phase == v1.PodUnknown {
+		df := func(opts *client.DeleteOptions) {
+			s := 5
+			opts.GracePeriodSeconds = &s // Delete the pod after 5 seconds.
+		}
+		err := r.client.Delete(ctx, pod, df)
```
Comment (hasbro17) [inline doc/user/client.md]: You can use the runtime client's helper function for this to make it simpler:
```Go
import ""sigs.k8s.io/controller-runtime/pkg/client""

err := r.client.Delete(ctx, pod, client.GracePeriodSeconds(5))

```

File: doc/user/client.md [hunk 11]
```diff
@@ -0,0 +1,411 @@
...
+func (r *ReconcileApp) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	...
+	
+	// Return all pods with a label of request.NamespacedName
+	opts := &client.ListOptions{}
+	opts.SetLabelSelector(request.NamespacedName)
+
+	pod := &v1.Pod{}
+	ctx := context.TODO()
+	err := r.client.Get(ctx, opts, pod)
```
Comment (hasbro17) [inline doc/user/client.md]: Change `Get` to `List` and use List object.
```Go
podList := &v1.PodList{}
err := r.client.Get(context.TODO(), opts, podList)
```

File: doc/user/client.md [hunk 12]
```diff
@@ -0,0 +1,411 @@
...
+}
+```
+
+```Go
+func (r *ReconcileApp) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	...
+	
+	// Return all pods with a label of request.NamespacedName
+	opts := &client.ListOptions{}
+	opts.SetLabelSelector(request.NamespacedName)
```
Comment (hasbro17) [inline doc/user/client.md]: [request.NamespacedName](https://github.com/kubernetes-sigs/controller-runtime/blob/master/vendor/k8s.io/apimachinery/pkg/types/namespacedname.go#L31-L34) is not the correct input to this helper.
This needs a raw string in the format of a label selector e.g `foo=bar` or `environment in (production, qa)` etc.
https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#label-selectors

We can demonstrate this better via:
```Go
// // Return all pods in the request namespace with a label of `app=<name>`
opts := &client.ListOptions{}
opts.SetLabelSelector(fmt.Sprintf("app=%s", request.NamespacedName.Name))
opts.InNamespace(request.NamespacedName.Namespace)
```

File: doc/user/client.md [hunk 13]
```diff
@@ -0,0 +1,411 @@
...
+```
+
+```Go
+func (r *ReconcileApp) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	...
+	
+	app := &v1alpha1.App{}
+	ctx := context.TODO()
+	name := request.NamespacedName
+	err := r.client.Get(ctx, name, app)
```
Comment (hasbro17) [inline doc/user/client.md]: Get takes [types.NamespacedName](https://github.com/kubernetes-sigs/controller-runtime/blob/master/vendor/k8s.io/apimachinery/pkg/types/namespacedname.go#L31-L34) as input. Not name.
```Go
err := r.client.Get(context.TODO(), request.NamespacedName, app)
```

And probably move the section for `Get` before `Delete` and `Update` since they're both using Get in their examples.

File: doc/user/client.md [hunk 14]
```diff
@@ -0,0 +1,411 @@
+# Operator-SDK: Controller Runtime Client API
+
+## Overview
+
+The [`controller-runtime`][repo-controller-runtime] library provides various abstractions to watch and reconcile resources in a k8s cluster via CRUD (Create, Update, Delete, as well as Get and List in this case) operations. Operators use at least one controller to perform a coherent set of tasks within a cluster, usually through a combination of CRUD operations. The Operator SDK uses controller-runtime's [Client][doc-client-client] interface, which defines a Reader and Writer to perform these operations.
```
Comment (hasbro17) [inline doc/user/client.md]: `which defines a Reader and Writer to perform these operations`
Since we don't really explain or go over the Reader and Writer interfaces, nor do we need to since they're internal interfaces, we can reword this to:
`which provides the interface for CRUD operations`

File: doc/user/client.md (hunk 13 above)
Comment (estroz) [reply doc/user/client.md]: `name` is `request.NamespacedName`, I moved it to its own variable for clarity.

Comment (hasbro17) [reply doc/user/client.md]: Yeah the variable `name` makes it seem like that parameter is just the name, when it's not. 
So maybe just pass it in directly. That should still be clear enough imo.

File: doc/user-guide.md [hunk 15]
```diff
@@ -156,7 +156,15 @@ return reconcile.Result{}, err
 return reconcile.Result{Requeue: true}, nil
 ```
 
-**// TODO:** Doc on controller-runtime client and examples on how to use it.
+You can set the `Resulte.RequeueAfter` to requeue the `Request` after a grace period as well:
```
Comment (LiliC) [inline doc/user-guide.md]:
```suggestion
You can set the `Result.RequeueAfter` to requeue the `Request` after a grace period as well:
```

Just had to try the new github suggestion :)

Comment (LiliC) [reply doc/user-guide.md]: Feel free to just change it instead of accepting that change. :)

File: doc/user/client.md [hunk 16]
```diff
@@ -0,0 +1,443 @@
+# Operator-SDK: Controller Runtime Client API
+
+## Overview
+
+The [`controller-runtime`][repo-controller-runtime] library provides various abstractions to watch and reconcile resources in a k8s cluster via CRUD (Create, Update, Delete, as well as Get and List in this case) operations. Operators use at least one controller to perform a coherent set of tasks within a cluster, usually through a combination of CRUD operations. The Operator SDK uses controller-runtime's [Client][doc-client-client] interface, which provides the interface for these operations.
```
Comment (LiliC) [inline doc/user/client.md]: Would prefer to spell out k8s in full,`s/k8s/kubernetes/ `.

File: doc/user/client.md [hunk 17]
```diff
@@ -0,0 +1,443 @@
...
+Creating a new Client is not usually necessary nor advised, as the default Client is sufficient for most use cases.
+
+### Reconcile and the Client API
+
+A Reconciler implements the [`reconcile.Reconciler`][doc-reconcile-reconciler] interface, which exposes the Reconcile method. Reconcilers are added to a corresponding Controller for a Kind; Reconcile is called in response to cluster or external Events, with a `reconcile.Request` object argument, to read and write cluster state by the Controller, and returns a `reconcile.Result`. SDK Reconcilers have access to a Client in order to make k8s API calls.
+
+**Note**: For those familiar with the SDK's old project semantics, [Handle][doc-osdk-handle] received resource events and reconciled state for multiple resource types, whereas Reconcile receives resource events and reconciles state for a single resource type.
+
+```Go
+// ReconcileMemcached reconciles a Kind object
```
Comment (joelanford) [inline doc/user/client.md]: Should this be `// ReconcileKind reconciles a Kind object`?

File: doc/user-guide.md [hunk 18]
```diff
@@ -156,7 +156,15 @@ return reconcile.Result{}, err
...
+```Go
+import "time"
+
+// Reconcile for any reason than error after 5 seconds
+return reconcile.Result{RequeueAfter: time.Duration(5)}, nil
```
Comment (shawn-hurley) [inline doc/user-guide.md]: nit:
```suggestion
return reconcile.Result{RequeueAfter: time.Second*5}, nil
```

 I prefer this way because it more clearly shows the unit IMO

Comment (AlexNPavel) [reply doc/user-guide.md]: I think `time.Duration(5)` would be 5 nanoseconds as well, not 5 seconds

File: doc/user-guide.md (hunk 15 above)
Comment (estroz) [reply doc/user-guide.md]: Woah very cool!

File: doc/user-guide.md [hunk 19]
```diff
@@ -156,7 +156,15 @@ return reconcile.Result{}, err
 return reconcile.Result{Requeue: true}, nil
 ```
 
-**// TODO:** Doc on controller-runtime client and examples on how to use it.
+You can set the `Result.RequeueAfter` to requeue the `Request` after a grace period as well:
+```Go
+import "time"
+
+// Reconcile for any reason than error after 5 seconds
+return reconcile.Result{RequeueAfter: time.Second*5}, nil
```
Comment (hasbro17) [inline doc/user-guide.md]: This should actually be `time.Duration(5)*time.Second` otherwise there's a type mismatch. time.Second is int64.

Comment (estroz) [reply doc/user-guide.md]: Seems to run fine as-is: https://play.golang.org/p/jex0ymMFT9A

File: doc/user/client.md [hunk 20]
```diff
@@ -0,0 +1,443 @@
...
+	client client.Client
+	scheme *runtime.Scheme
+}
+```
+
+A split client reads (Get and List) from the Cache and writes (Create, Update, Delete) to the API server. Reading from the Cache significantly reduces request load on the API server; as long as the Cache is updated by the API server, read operations are eventually consistent. 
+
+### Non-default Client
+
+An operator developer may wish to create their own Client that serves read requests(Get List) from the API server instead of the cache, for example. controller-runtime provides a [constructor][doc-client-constr] for Clients:
```
Comment (hasbro17) [inline doc/user/client.md]: Typo? There's a period in here.
`for example. controller-runtime provides`

File: doc/user/client.md [hunk 21]
```diff
@@ -0,0 +1,443 @@
...
+
+```Go
+// New returns a new Client using the provided config and Options.
+func New(config *rest.Config, options client.Options) (client.Client, error)
```
Comment (hasbro17) [inline doc/user/client.md]: We should probably give a small example of how to create a new client with the config. Otherwise the expectation might be the SDK provides the config(like it previously did with `sdk.GetKubeConfig()`).
```Go
import (
    "sigs.k8s.io/controller-runtime/pkg/client"
    "sigs.k8s.io/controller-runtime/pkg/client/config"
)

cfg, err := config.GetConfig()
...
c, err := client.New(cfg, client.Options{})
...
```

File: doc/user/client.md [hunk 22]
```diff
@@ -0,0 +1,443 @@
...
+		// Delete the pod after 5 seconds.
+		err := r.client.Delete(ctx, pod, client.GracePeriodSeconds(5))
+		...
+	}
+
+	...
+}
+```
+
+#### Get
```
Comment (hasbro17) [inline doc/user/client.md]: Nit but can we move this section before Update and Delete. Since both of those use Get before Get is explained.
